designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,950 @@
+1. An `@eslint/eslintrc` utility is provided to aid with backwards compatibility
+1. Remove the concept of environments (`env`) (including `--env`, `/*eslint-env*/`, and `env` in config files)
+1. Remove `.eslintignore` and `--ignore-file` (no longer necessary)
+1. Remove `--rulesdir`
+1. Remove `--ext`
+1. Remove `--resolve-plugins-relative-to`
+1. Replace `context.parserPath` with `context.parser` (the path to the parser is no longer valid)
+1. Allow plugins to specify a `parsers` key so parsers can be referenced the same as rules and processors in plugins.
+
+### The `eslint.config.js` File
+
+The `eslint.config.js` file is a JavaScript file (there is no JSON or YAML equivalent) that exports an object: 
+
+```js
+module.exports = {
+    name: "name",
+    files: ["*.js"],
+    ignores: ["*.test.js"],
+    settings: {},
+    languageOptions: {
+        ecmaVersion: 2020,
+        sourceType: "module",
+        globals: {},
+        parser: object || "string",
+        parserOptions: {},
+    }
+    linterOptions: {
+        reportUnusedDisableDirectives: "string"
+    },
+    processor: object || "string",
+    plugins: {}
+    rules: {}
+};
+```
+
+The following keys are new to the `eslint.config.js` format:
+
+* `name` - Specifies the name of the config object. This is helpful for printing out debugging information and, while not required, is recommended for that reason.
+* `files` - Determines the glob file patterns that this configuration applies to.
+* `ignores` - Determines the files that should not be linted using ESLint. This can be used in place of the `.eslintignore` file. The files specified by this array of glob patterns are subtracted from the files specified in `files`. If there is no `files` key, then `ignores` acts the same as `ignorePatterns` in `.eslintrc` files.
+
+The following keys are specified the same as in `.eslintrc` files:
+
+* `settings`
+* `rules`
+
+The following keys are specified differently than in `.eslintrc` files:
+
+* `plugins` - an object mapping plugin names to implementations. This replaces the `plugins` key in `.eslintrc` files and the `--rulesdir` option.
+* `processor` - an object or string in `eslint.config.js` files (a string in `.eslintrc`)
+* `languageOptions` - top-level grouping for all options that affect how JavaScript is interpreted by ESLint.
+    * `ecmaVersion` - sets the JavaScript version ESLint should use for these files. This value is copied into `parserOptions` if not already present.
+    * `sourceType` - sets the source type of the JavaScript code to parse. One of `module`, `script`, or `commonjs`.
+    * `parser` - an object or string in `eslint.config.js` files (a string in `.eslintrc`)
+    * `parserOptions` - an object specifying any additional parameters to be passed to the parser.
+    * `globals` - any additional global variables to add.
```

Comment:
I have curious, why `processor` is not a language option, differently from `parser`? (or why `parser` is a language option?)

- We may use the `parser` field to lint other than JavaScript. For example, `@typescript-eslint/parser` for TypeScript, `vue-eslint-parser` for Vue.js template files, and maybe for json, for graphql, etc.  
  `babel-eslint` may be a special case; it's a parser for JavaScript.
- We may use the `processor` field to lint other than JavaScript. For example, `eslint-plugin-html`, `eslint-plugin-markdown`, `eslint-plugin-svelte3`, etc.

We are using either field to parse other than JavaScript.

Reply:
Good point. I think it makes sense `processor` to be inside `languageOptions` too for the same reason `parser` is. It processes a language thing. Parser is language thing too.

Reply:
My thought was that processors don't really do anything that is language-specific. All they do is take text in one format and then split it out into a bunch of smaller pieces of text that can then be parsed. For example, `eslint-plugin-markdown` isn't really an option for JavaScript, it's an option for how to process the file, and then the language settings apply based on the virtual filenames being returned.

Reply:
Parsers are the same. We have used custom parsers to parse other languages than JavaScript. `babel-eslint` is the only exception that it parses JavaScript.

Reply:
The way I see it, `parser` is an option of the JavaScript language features built into ESLint, whereas `processor` has nothing JavaScript-specific in it. I'd like to leave this as-is, as I can anticipate extending configs in the future to be able to do stuff like this:

```js
[{
    files: ["*.css"],
    language: new CSSImplementation(),
    languageOptions: {
        something: true
    }
},
{
    files: ["*.js"],
    language: new JSImplementation(),
    languageOptions: {
        globals: {}
    }
}]
```

Right now I see our internal stuff as a JavaScript language implementation, and we might ultimately want to swap that out at some point.

Reply:
Hmm.

In the real world, there are multiple parsers that parse other languages than JavaScript and we have used those: `@typescript-eslint/parser`, `vue-eslint-parser`, `@graphql-eslint/parser`, etc. Therefore, because the `parser` is the setting that decides the language, I feel odd if `parser` is a JavaScript option.

Actually, all of `languageOptions` are parser options. The `languageOptions.ecmaVersion` and `languageOptions.sourceType` will be merged into `languageOptions.parserOptions`. The only exception is `languageOptions.globals`, but because the `parser` does scope analysis and returns the scope object, we can think it's a kind of parser options.

That said, I'm OK as-is because you are thinking the mechanism to switch the language that includes the parser and other parts.